### ISSUE TITLE
"repo update" fails on missing cache file

### ISSUE THREAD

**Issue by elouanKeryell-Even** [helm/helm#4659]
Output of `helm version`:

```
# helm version
Client: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
```

Output of `kubectl version`:

```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.1", GitCommit:"d4ab47518836c750f9949b9e0d387f20fb92260b", GitTreeState:"clean", BuildDate:"2018-04-12T14:26:04Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.8", GitCommit:"7eab6a49736cc7b01869a15f9f05dc5b49efb9fc", GitTreeState:"clean", BuildDate:"2018-09-14T15:54:20Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): Custom

# Issue

`helm repo update` complains about a missing cache file:

```
# helm repo update --debug
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Unable to get an update from the "stable" chart repository (https://kubernetes-charts.storage.googleapis.com):
	open /root/.helm/repository/cache/.helm/repository/cache/stable-index.yaml: no such file or directory
Update Complete. ⎈ Happy Helming!⎈
```

# Investigation

Performing `helm init` again does not seem to fix anything

```
# helm init --client-only
$HELM_HOME has been configured at /root/.helm.
Not installing Tiller due to 'client-only' flag having been set
Happy Helming!
```

Here are the logs of the initial client-side `helm init`:

```
# helm init --client-only --service-account tiller
Creating .helm
Creating .helm/repository
Creating .helm/repository/cache
Creating .helm/repository/local
Creating .helm/plugins
Creating .helm/starters
Creating .helm/cache/archive
Creating .helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
Adding local repo with URL: http://127.0.0.1:8879/charts
$HELM_HOME has been configured at .helm.
Not installing Tiller due to 'client-only' flag having been set
Happy Helming!
```

Here is the content of `$HELM_HOME`:

```
# find /root/.helm
/root/.helm
/root/.helm/repository
/root/.helm/repository/cache
/root/.helm/repository/cache/stable-index.yaml
/root/.helm/repository/cache/local-index.yaml
/root/.helm/repository/local
/root/.helm/repository/local/index.yaml
/root/.helm/repository/repositories.yaml
/root/.helm/plugins
/root/.helm/starters
/root/.helm/cache
/root/.helm/cache/archive
```

Content of `"$HELM_HOME/repository/repositories.yaml"`:

```
# cat /root/.helm/repository/repositories.yaml
apiVersion: v1
generated: 2018-09-18T15:41:35.279962679Z
repositories:
- caFile: ""
  cache: .helm/repository/cache/stable-index.yaml
  certFile: ""
  keyFile: ""
  name: stable
  password: ""
  url: https://kubernetes-charts.storage.googleapis.com
  username: ""
- caFile: ""
  cache: .helm/repository/cache/local-index.yaml
  certFile: ""
  keyFile: ""
  name: local
  password: ""
  url: http://127.0.0.1:8879/charts
  username: ""
```

# Potentially related issues

https://github.com/helm/helm/issues/2071

# Operational fix

Creating the missing directory structure seems to fix the error:

```
# mkdir -p "$HELM_HOME/repository/cache/.helm/repository/cache"
# helm repo update
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈ Happy Helming!⎈
```

**Comment by terribleplan** [comment]
I am also experiencing this. It seems to only happen on our CI runner, if I run the container the CI box executes everything works just fine. It almost looks like it is somehow getting confused when `HELM_HOME` is a relative path. Logs attached:

```
Running with gitlab-runner 11.1.0 (081978aa)
  on [redacted] b5838d45
Using Docker executor with image [redacted]:latest ...
Pulling docker image [redacted]:latest ...
Using docker image [redacted] for [redacted]:latest ...
Running on runner-b5838d45-project-2565-concurrent-0 via [redacted]...
Cloning repository...
Cloning into '[redacted]'...
Checking out [redacted] as master...
Skipping Git submodules setup
Downloading artifacts for [redacted]...
Downloading artifacts from coordinator... ok        id=75746 responseStatus=200 OK token=[redacted]
$ echo $HELM_HOME
.helm
$ rm -rf $HELM_HOME
$ mkdir $HELM_HOME
$ kubectl apply -f ./k8s/tiller.yml
serviceaccount/tiller unchanged
clusterrolebinding.rbac.authorization.k8s.io/tiller configured
$ helm --debug init --service-account tiller --wait || true
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      containers:
      - env:
        - name: TILLER_NAMESPACE
          value: kube-system
        - name: TILLER_HISTORY_MAX
          value: "0"
        image: gcr.io/kubernetes-helm/tiller:v2.10.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /liveness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        name: tiller
        ports:
        - containerPort: 44134
          name: tiller
        - containerPort: 44135
          name: http
        readinessProbe:
          httpGet:
            path: /readiness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        resources: {}
      serviceAccountName: tiller
status: {}

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  ports:
  - name: tiller
    port: 44134
    targetPort: tiller
  selector:
    app: helm
    name: tiller
  type: ClusterIP
status:
  loadBalancer: {}

...
Creating .helm/repository 
Creating .helm/repository/cache 
Creating .helm/repository/local 
Creating .helm/plugins 
Creating .helm/starters 
Creating .helm/cache/archive 
Creating .helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
Adding local repo with URL: http://127.0.0.1:8879/charts 
$HELM_HOME has been configured at .helm.
Warning: Tiller is already installed in the cluster.
(Use --client-only to suppress this message, or --upgrade to upgrade Tiller to the current version.)
[debug] Created tunnel using local port: '33617'

[debug] SERVER: "127.0.0.1:33617"

Happy Helming!
$ kubectl -n kube-system rollout status deployment tiller-deploy
deployment "tiller-deploy" successfully rolled out
$ helm --debug init --service-account tiller --upgrade --wait
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      containers:
      - env:
        - name: TILLER_NAMESPACE
          value: kube-system
        - name: TILLER_HISTORY_MAX
          value: "0"
        image: gcr.io/kubernetes-helm/tiller:v2.10.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /liveness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        name: tiller
        ports:
        - containerPort: 44134
          name: tiller
        - containerPort: 44135
          name: http
        readinessProbe:
          httpGet:
            path: /readiness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        resources: {}
      serviceAccountName: tiller
status: {}

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: kube-system
spec:
  ports:
  - name: tiller
    port: 44134
    targetPort: tiller
  selector:
    app: helm
    name: tiller
  type: ClusterIP
status:
  loadBalancer: {}

...
$HELM_HOME has been configured at .helm.
[debug] Created tunnel using local port: '45827'

[debug] SERVER: "127.0.0.1:45827"


Tiller (the Helm server-side component) has been upgraded to the current version.
[debug] SERVER: "127.0.0.1:45827"

Happy Helming!
$ kubectl -n kube-system rollout status deployment tiller-deploy
deployment "tiller-deploy" successfully rolled out
$ helm --debug repo update
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Unable to get an update from the "stable" chart repository (https://kubernetes-charts.storage.googleapis.com):
  open .helm/repository/cache/.helm/repository/cache/stable-index.yaml: no such file or directory
Update Complete. ⎈ Happy Helming!⎈ 
$ helm --debug repo add rancher-stable https://releases.rancher.com/server-charts/stable
Error: Looks like "https://releases.rancher.com/server-charts/stable" is not a valid chart repository or cannot be reached: open .helm/repository/cache/.helm/repository/cache/rancher-stable-index.yaml: no such file or directory
ERROR: Job failed: exit code 1
```

**Comment by terribleplan** [comment]
It looks like in my case it absolutely was related to how path resolution was working, once I set it to an absolute path it worked. I would consider this a better workaround to the one @elouanKeryell-Even mentioned in his original report.

```
$ echo $HELM_HOME
.helm
$ rm -rf $HELM_HOME
$ mkdir $HELM_HOME
$ export HELM_HOME=$(cd $HELM_HOME && pwd)
$ echo $HELM_HOME
/builds/[redacted]/.helm
[snip]
$ helm --debug repo update
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈ Happy Helming!⎈ 
$ helm --debug repo add rancher-stable https://releases.rancher.com/server-charts/stable
"rancher-stable" has been added to your repositories
Uploading artifacts...
.helm: found 13 matching files                     
Uploading artifacts to coordinator... ok            [redacted]
[redacted]
Job succeeded
```

**Comment by davidfurlong** [comment]
Reinstalling helm fixed this for me

**Comment by huang-jy** [comment]
Adding my own experience here.

In my case, this error occurred because I copied my home directory off in preparation for a clean reinstall of my OS, and then copied it back.

What I didn't realise is that `~/.helm/repository/repositories.yaml` had my old home directory name (my username changed between the old and new OS installation) and this broke my "local" repo.

Doing a simple `sed -i s/old-user-name/new-user-name/p ~/.helm/repository/repositories.yaml` fixed it in my case.

**Comment by kivagant-ba** [comment]
The workaround from @terribleplan adopted for Jenkins:
```
def helmHome = sh(returnStdout: true, script: 'rm -rf .helm/ && mkdir .helm && cd .helm && pwd').trim()
            withEnv(["HELM_HOME=${helmHome}"]) {
...
}
```

**Comment by maxh8086** [comment]
Hello team,

i am getting below error 

Please guide 

helm  repo add rancher-stable https://releases.rancher.com/server-charts/stable
Error: Looks like "https://releases.rancher.com/server-charts/stable" is not a valid chart repository or cannot be reached: Get https://releases.rancher.com/server-charts/stable/index.yaml: Forbidden

**Comment by Davidiusdadi** [comment]
TL;DR **specifying the HELM_HOME as an absolute path resolved the error for me**

i got the same error message and could fix it by changing my `HELM_HOME`
it was set to `HELM_HOME="~/.helm_X"` and i had to change it to  `HELM_HOME="/home/user/.helm_X"`. <br> The `~` was not resolved to my home-dir but instead `helm init` created a folder called `~` in my home-dir... XD

I am aware that the issue-creators path is an absolute one - but since i got the same error message i thought i share my solution anyway.

**Comment by rbutola** [comment]
I also faced same issue. My test system was launching in a fresh container which did not have HOME variable set. This led to HELM_HOME to be empty. This caused helm installation to fail.
Workaround mentioned by @terribleplan worked fine:
In golang:
        _ = os.Mkdir("./.helm", 0700)
        dir, _ := os.Getwd()
        os.Setenv("HELM_HOME", dir+"/"+".helm")

**Comment by Jyoti492** [comment]
Hi Team,
I am also facing the same issue but while fetching the helm charts from the ACR.
Error: Looks like "https://mcpshareddcr.azurecr.io/helm/v1/repo" is not a valid chart repository or cannot be reached: open .helm/repository/cache/.helm/repository/cache/mcpshareddcr-index.yaml: no such file or directory

followed all above but that just adds local repo not ACR

**Comment by bacongobbler** [comment]
We are no longer accepting bug fixes for Helm 2. See https://helm.sh/blog/helm-v2-deprecation-timeline/ for more details.

**Comment by chance2021** [comment]
I have just updated **requirements.yaml** file with the repository link to **https://charts.helm.sh/stable *, and then I run` helm dependency update` and it works for me
```
xwiki/xwiki-helm$ cat requirements.yaml 
dependencies:
  - name: mysql
    #repository: https://kubernetes-charts.storage.googleapis.com
    repository: https://charts.helm.sh/stable
    version: 1.4.0
    condition: mysql.enabled
  - name: postgresql
    #repository: https://kubernetes-charts.storage.googleapis.com
    repository: https://charts.helm.sh/stable
    version: 7.6.0
    condition: postgresql.enabled
```
